### PR TITLE
feat: onIncoming and onTokenWillExpire

### PR DESCRIPTION
**Pull request by kpchoy**
[IDL](https://docs.google.com/document/d/1rHlTs9y7TcNZLaHx2aHJAgo7tEwqyvn-Izbf0vkoLYQ/edit)
[VBLOCKS-3463](https://twilio-engineering.atlassian.net/browse/VBLOCKS-3463)

- onTokenWillExpire (EventHandler)
- onIncoming (EventHandler)

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: src/components/twilio-voice-dialer/public/index.js [hunk 1]
```diff
@@ -4,4 +4,16 @@
...
+    const device = e.detail.device;
+    console.log('onTokenWillExpire', device);
+
+    const updateTokenResponse = await fetch(`/twilio-voice-dialer/token`);
+    const updateTokenData = await updateTokenResponse.json();
+    device.updateToken(updateTokenData.token);
+  });
+  twilioVoiceDialer.addEventListener('onIncoming', (e) => {
+    const call = e.detail.call;
+    console.log('onIncoming', call);
```
Comment (kpchoy) [inline src/components/twilio-voice-dialer/public/index.js:17]: Not sure the ideal example to put here, any suggestions?

File: src/components/twilio-voice-dialer/public/index.js [hunk 2]
```diff
@@ -4,4 +4,16 @@
 
   const twilioVoiceDialer = document.querySelector('twilio-voice-dialer');
   twilioVoiceDialer.setToken(data.token);
+  twilioVoiceDialer.addEventListener('onTokenWillExpire', async (e) => {
+    const device = e.detail.device;
+    console.log('onTokenWillExpire', device);
```
Comment (mhuynh5757) [inline src/components/twilio-voice-dialer/public/index.js:9]: We might not want to log the `device` here, might clutter the console. WDYT?

File: src/components/twilio-voice-dialer/public/index.js (hunk 1 above)
Comment (mhuynh5757) [reply src/components/twilio-voice-dialer/public/index.js:17]: Maybe just a comment saying something like "you can handle call events here" or even handling a call event like `call.on('disconnect', ...)`.

Comment (mhuynh5757) [reply src/components/twilio-voice-dialer/public/index.js:17]: Is this component supposed to have full incoming and outgoing call functionality, or just outgoing?

File: src/components/twilio-voice-dialer/public/index.js
Comment (mhuynh5757) [inline src/components/twilio-voice-dialer/public/index.js:1]: For our event names, we might want to consider dropping the `on-` prefix. For example, `onTokenWillExpire` should be `tokenWillExpire`. If you think about it, the method `.addEventListener` has the synonym `.on` with most event libraries. And then we would have `.on('onTokenWillExpire', ...)` which feels clunky.

File: src/components/twilio-voice-dialer/public/index.js (hunk 1 above)
Comment (kpchoy) [reply src/components/twilio-voice-dialer/public/index.js:17]: The component has both functionality. However we only expose the incoming event. See IDL: https://docs.google.com/document/d/1rHlTs9y7TcNZLaHx2aHJAgo7tEwqyvn-Izbf0vkoLYQ/edit#heading=h.xf1ox0g8pcv6

File: src/components/twilio-voice-dialer/public/index.js
Comment (kpchoy) [reply src/components/twilio-voice-dialer/public/index.js:1]: that is true. Ok will change the naming for the events to drop the `on` prefix!

Comment (kpchoy) [reply src/components/twilio-voice-dialer/public/index.js:1]: @mhuynh5757 Do you think I should prefix with `event-`?
`eventTokenWillExpire` and `eventIncoming` or just `tokenWillExpire` and `incoming`

Comment (mhuynh5757) [reply src/components/twilio-voice-dialer/public/index.js:1]: Just `tokenWillExpire` and `incoming`, like how the Voice JS SDK does it.

File: src/components/twilio-voice-dialer/public/twilio-voice-dialer.js [hunk 3]
```diff
@@ -96,7 +96,7 @@ class TwilioVoiceDialer extends HTMLElement {
   }
 
   #incomingHandler = (call) => {
```
Comment (mhuynh5757) [inline src/components/twilio-voice-dialer/public/twilio-voice-dialer.js:98]: I think we should rename some of the methods. Just from names, it's hard to tell what `incomingHandler` and `incomingEvent` do. The other handlers are labeled `handleX`, should we do the same with `incomingHandler` and rename to `handleIncoming`? For the `incomingEvent` and `tokenWillExpireEvent`, those might be better named as `raiseIncomingEvent` and `raiseTokenWillExpireEvent`? I think methods being verbs is a good idea.

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]

File: src/components/twilio-voice-dialer/public/twilio-voice-dialer.js (hunk 3 above)
Comment (kpchoy) [reply src/components/twilio-voice-dialer/public/twilio-voice-dialer.js]: I like this idea 👍

File: src/components/twilio-voice-dialer/public/twilio-voice-dialer.js [hunk 4]
```diff
@@ -95,28 +105,18 @@ class TwilioVoiceDialer extends HTMLElement {
...
-    call.on('disconnect', this.#reset);
-    call.on('cancel', this.#reset);
-    call.on('reject', this.#reset);
-    this.#call = call;
-    this.#setStatus('incoming');
-  };
-
-  #incomingEvent(call) {
-    const incomingEvent = new CustomEvent('incoming', {
+  #raiseIncomingEvent(call) {
```
Comment (mhuynh5757) [inline src/components/twilio-voice-dialer/public/twilio-voice-dialer.js:108]: Sorry for being so nitpicky: I see that we actually "dispatch" events, should we rename `raise` to `dispatch` instead? I didn't notice sooner, apologies.

Also, the events that we're creating with `new CustomEvent` should not have the labels `raiseX`, just `X`. Like on line 109, the variable should just be `incomingEvent` and not `raiseIncomingEvent`.

**Review by mhuynh5757 — CHANGES_REQUESTED** [review]

File: src/components/twilio-voice-dialer/public/twilio-voice-dialer.js (hunk 4 above)
Comment (kpchoy) [reply src/components/twilio-voice-dialer/public/twilio-voice-dialer.js]: no worries! agreed `dispatch` will match

nice catch. yes makes sense for CustomEvent to drop the prefix.

**Review by mhuynh5757 — APPROVED** [review]

**Pull request merged**